pkg/ruler/ruler.go

```diff
@@ -903,12 +1012,19 @@ func (r *Ruler) getShardedRules(ctx context.Context, userID string, rulesRequest
 	}
 
 	var (
-		mergedMx sync.Mutex
-		merged   []*GroupStateDesc
+		mtx    sync.Mutex
+		merged []*GroupStateDesc
+		errs   []error
 	)
+	failedZones := make(map[string]interface{})
```

Comment:
I would recommend using  `make(map[string]struct{})` which is used more common in go to represent a set.
IIUC this is what we need.